testing/trino-tests/src/test/java/io/trino/execution/EventsCollector.java

```diff
@@ -14,171 +14,174 @@
     }
 
-    public synchronized List<QueryCreatedEvent> getQueryCreatedEvents()
+    public QueryEvents getQueryEvents(QueryId queryId)
     {
-        return queryCreatedEvents.build();
+        return queryEvents.computeIfAbsent(queryId, ignored -> new QueryEvents());
     }
 
-    public synchronized List<QueryCompletedEvent> getQueryCompletedEvents()
+    @ThreadSafe
+    public static class QueryEvents
     {
-        return queryCompletedEvents.build();
-    }
+        @GuardedBy("this")
+        private QueryCreatedEvent queryCreatedEvent;
+        @GuardedBy("this")
+        private QueryCompletedEvent queryCompletedEvent;
+        @GuardedBy("this")
+        private final CountDownLatch queryCompleteLatch = new CountDownLatch(1);
+
+        @GuardedBy("this")
+        private final List<SplitCompletedEvent> splitCompletedEvents = new ArrayList<>();
+        @GuardedBy("this")
+        private CountDownLatch splitEventLatch;
+
+        @GuardedBy("this")
+        private final List<Exception> failures = new ArrayList<>();
+
+        public synchronized QueryCreatedEvent getQueryCreatedEvent()
+        {
+            checkFailure();
+            if (queryCreatedEvent == null) {
+                throw new IllegalStateException("QueryCreatedEvent has not been set");
+            }
+            return queryCreatedEvent;
+        }
 
-    public synchronized List<SplitCompletedEvent> getSplitCompletedEvents()
-    {
-        return splitCompletedEvents.build();
-    }
+        public synchronized QueryCompletedEvent getQueryCompletedEvent()
+        {
+            checkFailure();
+            if (queryCompletedEvent == null) {
+                throw new IllegalStateException("QueryCompletedEvent has not been set");
+            }
+            return queryCompletedEvent;
+        }
 
-    @ThreadSafe
-    public static class EventFilters
-    {
-        private final Predicate<QueryCreatedEvent> queryCreatedFilter;
-        private final Predicate<QueryCompletedEvent> queryCompletedFilter;
-        private final Predicate<SplitCompletedEvent> splitCompletedFilter;
-
-        private EventFilters(
-                Predicate<QueryCreatedEvent> queryCreatedFilter,
-                Predicate<QueryCompletedEvent> queryCompletedFilter,
-                Predicate<SplitCompletedEvent> splitCompletedFilter)
+        private synchronized void addQueryCreated(QueryCreatedEvent event)
         {
-            this.queryCreatedFilter = requireNonNull(queryCreatedFilter, "queryCreatedFilter is null");
-            this.queryCompletedFilter = requireNonNull(queryCompletedFilter, "queryCompletedFilter is null");
-            this.splitCompletedFilter = requireNonNull(splitCompletedFilter, "splitCompletedFilter is null");
+            if (queryCreatedEvent != null) {
+                RuntimeException exception = new RuntimeException("QueryCreateEvent already set");
+                exception.fillInStackTrace();
```

Comment:
What would that change? It is called in constuctor already.

```java
    public static void main(String[] args)
    {
        RuntimeException e1 = new RuntimeException("e1");
        RuntimeException e2 = new RuntimeException("e2");
        e2.fillInStackTrace();

        e1.printStackTrace();
        e2.printStackTrace();
    }
```
produces same stacktrace for both exceptions.
```
java.lang.RuntimeException: e1
	at io.trino.execution.EventsCollector.main(EventsCollector.java:190)
java.lang.RuntimeException: e2
	at io.trino.execution.EventsCollector.main(EventsCollector.java:192)
```